src/main/java/net/minecraftforge/common/ForgeConfigSpec.java

```diff
@@ -482,8 +505,24 @@ public <V extends Enum<V>> EnumValue<V> defineEnum(List<String> path, Supplier<V
         public <V extends Enum<V>> EnumValue<V> defineEnum(List<String> path, Supplier<V> defaultSupplier, EnumGetMethod converter, Predicate<Object> validator, Class<V> clazz) {
             context.setClazz(clazz);
             V[] allowedValues = clazz.getEnumConstants();
+            if (context.getErrorDescriber() == null) {
+                context.setErrorDescriber((obj) -> {
+                    if (obj instanceof String string) {
+                        return Component.translatable("forge.configgui.error.enum.invalidName", string, Arrays.stream(allowedValues).filter(validator).map(Enum::name).collect(Collectors.joining("\n - ", "\n - ", "")));
+                    }
+
+                    return Component.translatable("forge.configgui.error.enum.needsToBeText");
+                });
+            }
             context.setComment(ObjectArrays.concat(context.getComment(), "Allowed Values: " + Arrays.stream(allowedValues).filter(validator).map(Enum::name).collect(Collectors.joining(", "))));
-            return new EnumValue<V>(this, define(path, new ValueSpec(defaultSupplier, validator, context), defaultSupplier).getPath(), defaultSupplier, converter, clazz);
+            final EnumValue<V> result = new EnumValue<V>(this, define(path, new ValueSpec(defaultSupplier, validator, context), defaultSupplier).getPath(), defaultSupplier, converter, clazz);
+
+            if (this.widgetFactorySupplier != null) {
+                result.setScreenWidgetFactorySupplier(this.widgetFactorySupplier);
+                this.widgetFactorySupplier = null;
+            }
```

Comment:
With this block of code basically everywhere wouldn't it be a bit simpler as the constructors are package private (so not public api) to just add an extra nullable parameter to them that takes a nullable widget factory supplier? And then when calling the super constructor can pass their fallback version if this is null. And then the field that stores it could even be final.

Reply:
Because that would cause a gigantic mess of backward-compatible constructor creation.

This is a simple check to see if it is set, and yes it is a bit of a duplication but creating 6 new constructors and the relevant hooks is just as much of a mess.